open_flamingo/eval/evaluate.py

```diff
@@ -670,93 +683,158 @@ def evaluate_imagenet(
+                        clear_conditioned_layers=False,
+                        use_cached_vision_x=True,
+                        past_key_values=(
+                            past_key_values if token_idx > 0 else precomputed_pkvs
+                        ),
+                        use_cache=True,
+                    )
+                past_key_values = _detach_pkvs(outputs.past_key_values)
+                elementwise_logits.append(outputs.logits.detach())
+
+            # logits/probs has shape [B, classname_tokens + 1, vocab_size]
+            logits = torch.concat(elementwise_logits, 1)
+            probs = torch.softmax(logits, dim=-1).detach()
 
-            outputs = model(
-                vision_x=None,
-                lang_x=lang_x["input_ids"].cuda(),
-                attention_mask=lang_x["attention_mask"].cuda(),
-                clear_conditioned_layers=False,
-                use_cached_vision_x=True,
-            )
-            probs = torch.softmax(outputs.logits, dim=-1).detach()
             # collect the probability of the generated token -- probability
-            # at index 0 corresponds to the token at index 1
-            probs = probs[:, :-1, :]
-            input_ids = lang_x["input_ids"][:, 1:].cuda()
-            gen_probs = torch.gather(probs, 2, input_ids[:, :, None]).squeeze(-1)
-
-            probs = []
-            for input_sentence, input_probs in zip(input_ids, gen_probs):
-                idxes = find_sub_list(
-                    prompt_tokens, input_sentence.detach().cpu().numpy().tolist()
-                )
-                input_probs = input_probs[idxes[-1] + 1 :]
-                probs.append(torch.prod(input_probs).item())
-            overall_probs.append(probs)
+            # at index 0 corresponds to the token at index 1.
+            probs = probs[:, :-1, :]  # shape [B, classname_tokens, vocab_size]
 
-        top5 = [
-            IMAGENET_1K_CLASS_ID_TO_LABEL[pred]
-            for pred in np.argsort(np.array(overall_probs)[:, 0])[::-1][:5]
-        ]
-        if batch["class_name"] == top5[0]:
-            acc1 += 1
-        if batch["class_name"] in top5:
-            acc5 += 1
+            gen_probs = torch.gather(probs, 2, classname_tokens[:, :, None]).squeeze(-1)
+
+            class_prob = torch.prod(gen_probs, 1).detach().cpu().numpy()
+            overall_probs.append(class_prob)
+
+        overall_probs = np.row_stack(overall_probs).T  # shape [B, num_classes]
+
+        def topk(probs_ary: np.ndarray, k: int) -> np.ndarray:
+            """Return the indices of the top k elements in probs_ary."""
+            return np.argsort(probs_ary)[::-1][:k]
+
+        for i in range(batch_size):
+            top5 = [
+                IMAGENET_1K_CLASS_ID_TO_LABEL[pred]
+                for pred in topk(overall_probs[i], 5)
+            ]
+
+            y_i = batch[i]["class_name"]
+            acc5 += int(y_i in set(top5))
+            acc1 += int(y_i == top5[0])
+
+            print(
```

Comment:
Is this for debugging purposes? Should we remove?

Reply:
it is for debugging -- but I'd vote to leave it in for now. It's really useful to have a second data point for debugging, and sometimes top-5 is more informative than top-1 (i.e. it tends to have more true positives/be nonzero more, which is useful when doing quick checks to see that the metrics don't degrade to zero). 

This will be factored out in a future update though where we'll provide more flexibility over metrics for other datasets.